src/main/drivers/pwm_output_dshot_hal.c

```diff
@@ -36,63 +38,86 @@
 
 #include "pwm_output.h"
 
-typedef DMA_Stream_TypeDef dmaStream_t;
+// TODO remove once debugging no longer needed
+#ifdef USE_DSHOT_TELEMETRY
+#include <string.h>
+#endif
+
+#include "pwm_output_dshot_shared.h"
 
-static FAST_RAM_ZERO_INIT uint8_t dmaMotorTimerCount = 0;
-static FAST_RAM_ZERO_INIT motorDmaTimer_t dmaMotorTimers[MAX_DMA_TIMERS];
-static FAST_RAM_ZERO_INIT motorDmaOutput_t dmaMotors[MAX_SUPPORTED_MOTORS];
+#ifdef USE_DSHOT_TELEMETRY
 
-motorDmaOutput_t *getMotorDmaOutput(uint8_t index)
+static void processInputIrq(motorDmaOutput_t * const motor)
 {
-    return &dmaMotors[index];
+    motor->hasTelemetry = true;
+
+#ifdef USE_DSHOT_DMAR
+    if (useBurstDshot) {
+        LL_EX_DMA_DisableStream(motor->timerHardware->dmaTimUPRef);
+        LL_TIM_DisableDMAReq_UPDATE(motor->timerHardware->tim);
+    } else
+#endif
+    {
+        LL_EX_DMA_DisableStream(motor->dmaRef);
+        LL_EX_TIM_DisableIT(motor->timerHardware->tim, motor->timerDmaSource);
+    }
+    readDoneCount++;
 }
 
-uint8_t getTimerIndex(TIM_TypeDef *timer)
+static void enableChannels(uint8_t motorCount)
 {
-    for (int i = 0; i < dmaMotorTimerCount; i++) {
-        if (dmaMotorTimers[i].timer == timer) {
-            return i;
+    for (int i = 0; i < motorCount; i++) {
+        if (dmaMotors[i].output & TIMER_OUTPUT_N_CHANNEL) {
+            LL_EX_TIM_CC_EnableNChannel(dmaMotors[i].timerHardware->tim, dmaMotors[i].llChannel);
+        } else {
+            LL_TIM_CC_EnableChannel(dmaMotors[i].timerHardware->tim, dmaMotors[i].llChannel);
         }
     }
-    dmaMotorTimers[dmaMotorTimerCount++].timer = timer;
-    return dmaMotorTimerCount - 1;
 }
 
-FAST_CODE void pwmWriteDshotInt(uint8_t index, uint16_t value)
-{
-    motorDmaOutput_t *const motor = &dmaMotors[index];
+#endif
 
-    if (!motor->configured) {
-        return;
-    }
 
-    /*If there is a command ready to go overwrite the value and send that instead*/
-    if (pwmDshotCommandIsProcessing()) {
-        value = pwmGetDshotCommand(index);
-        if (value) {
-            motor->requestTelemetry = true;
-        }
-    }
+static void motor_DMA_IRQHandler(dmaChannelDescriptor_t *descriptor);
 
-    motor->value = value;
+inline static void pwmDshotSetDirectionOutput(
+    motorDmaOutput_t * const motor, bool output
+#ifndef USE_DSHOT_TELEMETRY
```

Comment:
Not sure if making function signatures conditional is a good approach - keeping the signature constant, and stubbing out in cases where some parts aren't used will be easier to maintain.

Reply:
This was just take from the original f4 implementation. I can change it in both. Originally I didn't want to add any overhead for implementations that don't USE_DSHOT_TELEMETRY.

Reply:
Not sure about the overhead. My assumption was that with LTO, the linker would take care of eliminating unused parameters from function calls. And if the function call is inlined (which we promote aggressively) then having additional unused parameters isn't going to make a difference anyway.
What do you guys think, @jflyper, @etracer65, @ledvinap?